### PR TITLE
Feature - SQL engine cache

### PR DESCRIPTION
**Pull request by khaitruong922**

### CODE REVIEW

File: src/taipy/core/data/abstract_sql.py [hunk 1]
```diff
@@ -28,17 +28,43 @@
...
+    __DB_USERNAME_KEY = "db_username"
+    __DB_PASSWORD_KEY = "db_password"
+    __DB_HOST_KEY = "db_host"
+    __DB_PORT_KEY = "db_port"
+    __DB_ENGINE_KEY = "db_engine"
+    __DB_DRIVER_KEY = "db_driver"
     __DB_EXTRA_ARGS_KEY = "db_extra_args"
+    __SQLITE_PATH_KEY = "path"
+
+    __ENGINE_PROERTIES: List[str] = [
```
Comment (jrobinAV) [inline src/taipy/core/data/abstract_sql.py]:
```suggestion
    __ENGINE_PROPERTIES: List[str] = [
```

File: src/taipy/core/data/abstract_sql.py [hunk 2]
```diff
@@ -73,26 +99,14 @@ def __init__(
             edit_in_progress,
             **properties,
         )
+        self.engine = None
```
Comment (jrobinAV) [inline src/taipy/core/data/abstract_sql.py]:
```suggestion
        self._engine = None
```

File: src/taipy/core/data/abstract_sql.py [hunk 3]
```diff
@@ -108,29 +122,37 @@ def _check_exposed_type(self, exposed_type):
...
-    @staticmethod
-    def __build_conn_string(
-        engine, username, host, password, database, port, driver, extra_args: Dict[str, str], path
-    ) -> str:
-        # TODO: Add support to other SQL engines, the engine value should be checked.
-        if engine == "mssql":
+    def _get_engine(self):
+        if self.engine is None:
+            self.engine = create_engine(self._conn_string())
+        return self.engine
```
Comment (jrobinAV) [inline src/taipy/core/data/abstract_sql.py]:
```suggestion
        if self._engine is None:
            self._engine = create_engine(self._conn_string())
        return self._engine
```

File: src/taipy/core/data/abstract_sql.py [hunk 4]
```diff
@@ -177,3 +199,8 @@ def _write(self, data) -> None:
                     raise
                 else:
                     transaction.commit()
+
+    def __setattr__(self, key: str, value) -> None:
+        if key in self.__ENGINE_PROERTIES:
+            self.engine = None
```
Comment (jrobinAV) [inline src/taipy/core/data/abstract_sql.py]:
```suggestion
            self._engine = None
```

File: src/taipy/core/data/abstract_sql.py [hunk 5]
```diff
@@ -177,3 +199,8 @@ def _write(self, data) -> None:
                     raise
                 else:
                     transaction.commit()
+
+    def __setattr__(self, key: str, value) -> None:
+        if key in self.__ENGINE_PROERTIES:
```
Comment (jrobinAV) [inline src/taipy/core/data/abstract_sql.py]:
```suggestion
        if key in self.__ENGINE_PROPERTIES:
```

File: tests/core/data/test_sql_table_data_node.py [hunk 6]
```diff
@@ -309,6 +309,40 @@ def test_write_empty_list(self, data):
...
+            properties={
+                "db_username": "sa",
+                "db_password": "foobar",
+                "db_name": "datanode",
+                "db_engine": "mssql",
+                "table_name": "foo",
+            },
+        )
+
+        assert dn.engine is None
```
Comment (jrobinAV) [inline tests/core/data/test_sql_table_data_node.py]:
```suggestion
        assert dn._engine is None
```

File: tests/core/data/test_sql_table_data_node.py [hunk 7]
```diff
@@ -309,6 +309,40 @@ def test_write_empty_list(self, data):
...
+
+        with mock.patch("sqlalchemy.engine.Engine.connect") as engine_mock, mock.patch(
+            "src.taipy.core.data.sql_table.SQLTableDataNode._create_table"
+        ):
+            cursor_mock = engine_mock.return_value.__enter__.return_value
+            cursor_mock.execute.side_effect = None
+
+            dn.read()
+            assert dn.engine is not None
```
Comment (jrobinAV) [inline tests/core/data/test_sql_table_data_node.py]:
```suggestion
            assert dn._engine is not None
```

File: tests/core/data/test_sql_table_data_node.py [hunk 8]
```diff
@@ -309,6 +309,40 @@ def test_write_empty_list(self, data):
...
+
+            dn.db_username = "foo"
+            assert dn.engine is None
```
Comment (jrobinAV) [inline tests/core/data/test_sql_table_data_node.py]:
```suggestion
            assert dn._engine is None
```

File: tests/core/data/test_sql_table_data_node.py [hunk 9]
```diff
@@ -309,6 +309,40 @@ def test_write_empty_list(self, data):
...
+
+            dn.write(1)
+            assert dn.engine is not None
```
Comment (jrobinAV) [inline tests/core/data/test_sql_table_data_node.py]:
```suggestion
            assert dn._engine is not None
```

File: tests/core/data/test_sql_table_data_node.py [hunk 10]
```diff
@@ -309,6 +309,40 @@ def test_write_empty_list(self, data):
...
+
+            dn.some_random_attribute_that_does_not_related_to_engine = "foo"
+            assert dn.engine is not None
```
Comment (jrobinAV) [inline tests/core/data/test_sql_table_data_node.py]:
```suggestion
            assert dn._engine is not None
```

**Review by jrobinAV — DISMISSED** [review]
This is great!

I just proposed to rename self.engine into self._engine.
It should NOT be a public attribute.

**Review by jrobinAV — COMMENTED** [review]
This PR should be part of core 2.1. We should wait for 2.0 to be released before merging it in develop.

**Review by joaoandre-avaiga — DISMISSED** [review]

**Review by toan-quach — DISMISSED** [review]

**Review by jrobinAV — DISMISSED** [review]

**Review by jrobinAV — APPROVED** [review]

**Pull request merged**